### ISSUE TITLE
More testing

### ISSUE THREAD

**Issue by bjorn3** [bjorn3/rustc_codegen_cranelift#247]
* [x] Run https://github.com/dropbox/rust-brotli-decompressor
* [x] Run https://github.com/Peltoche/lsd (<strike>hashbrown assertion failure due to #348, because debug assertions are disabled</strike> works as of a24a956aa6377e35633c39d2f134d642bde4bb25)
* [x] Run https://github.com/sitkevij/hex.git
* [x] Run https://github.com/sharkdp/hexyl (<strike>works when patching ctrl+c handler, which uses threads, away</strike> (#784))
* [x] Run https://github.com/ogham/exa (<strike>requires threads</strike> (#784))
* [x] Test https://github.com/rust-lang/regex (<strike>requires cpuid intrinsic</strike> (faked no cpuid support by cpu) and <strike>simd</strike> (implemented enough in #650))
* [x] Run https://github.com/mohanson/gameboy
* [ ] Run https://github.com/ozkriff/zemeroth (<strike>blocked on #666</strike>, requires threads)
* [x] Run https://gitlab.com/veloren/veloren (<strike>requires threads</strike> (#784), kind of works, but too slow to be usable)
* [x] Run https://github.com/ozkriff/zoc
* [x] Run https://github.com/gchp/iota
* [x] Test https://github.com/BurntSushi/fst
* [x] Run https://github.com/dalance/procs (<strike>requires `saturating_{add,sub}` for signed integers</strike> (3fcd54088c71cf64d723fe928f3da188f6d9952e))
* [x] Run https://github.com/nical/lyon examples (<strike>requires `saturating_{add,sub}` for signed integers</strike> (3fcd54088c71cf64d723fe928f3da188f6d9952e), all but the wgpu examples tested, my graphics driver doesn't have vulkan/metal/dx* support)
* [ ] Run more applications

**Comment by bjorn3** [comment]
If you know of a rust program which doesn't use threads, please let me know, so I can test compiling it using cg_clif.

**Comment by memoryruins** [comment]
- simple-raytracer https://github.com/ebobby/simple-raytracer
- svgbob https://github.com/ivanceras/svgbob
- svgcleaner https://github.com/RazrFalcon/svgcleaner
- bat https://github.com/sharkdp/bat
- fd https://github.com/sharkdp/fd (only creates a job server if a flag is passed)
- fselect https://github.com/jhspetersson/fselect
- encrusted https://github.com/DeMille/encrusted
- termpix https://github.com/hopey-dishwasher/termpix

(on mobile and unable to confirm if all of these don't use threads)

**Comment by bjorn3** [comment]
Thanks @memoryruins! Will try them.

**Comment by bjorn3** [comment]
* [ ] https://github.com/ebobby/simple-raytracer (blocked on #668)

**Comment by bjorn3** [comment]
@memoryruins Got all but the thread requiring programs working 🎉

**Comment by jyn514** [comment]
c2-chacha doesn't compile: `error: at fn core::core_arch[0]::x86_64[0]::sse41[0]::_mm_insert_epi64[0]: unsupported intrinsic simd_insert`

I also got `warning: err opening clif file: Os { code: 36, kind: Other, message: "File name too long" }`, not sure if that affects the build.

Everything else for https://github.com/jyn514/rcc works well, great work!

**Comment by jyn514** [comment]
Other misc projects:

- [ ] https://github.com/hsivonen/encoding_rs (used in cargo-deadlinks)
```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `2816`,
 right: `0`: faerie doesn't support addends in data section relocations yet', /home/joshua/.cargo/git/checkouts/cranelift-af8713b52459e6cd/01201b1/cranelift-faerie/src/backend.rs:245:13
```

**Comment by bjorn3** [comment]
Thanks for testing.

> https://crates.io/crates/c2-chacha doesn't compile (used in rand): error: at fn core::core_arch[0]::x86_64[0]::sse41[0]::_mm_insert_epi64[0]: unsupported intrinsic simd_insert

That ishttps://github.com/bjorn3/rustc_codegen_cranelift/issues/669. You can edit the Cargo.toml of c2-chacha to not passthrough the simd feature to one of its dependencies as workaround.

> I also got warning: err opening clif file: Os { code: 36, kind: Other, message: "File name too long" }, not sure if that affects the build.

No it doesnt. In debug mode cg_clif tries to write the clif ir of every compiled function to target/out/clif in the source dir for cg_clif. If the filename is too big, it just ignores it. You can build cg_clif in release mode to prevent writing of clif ir. That also makes it a lot faster.

>  https://github.com/hsivonen/encoding_rs (used in reqwest)

Known issue. faerie (the object file writer) doesnt implement something.

**Comment by bjorn3** [comment]
* [ ] Run https://github.com/BurntSushi/ripgrep
* [ ] Run https://github.com/m4b/bingrep
* [ ] Run https://github.com/gahag/bgrep

**Comment by jyn514** [comment]
`bingrep` works on my machine, not sure what error it's giving on yours.

```sh
$ uname -a
Linux joshua-ThinkPad 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by bjorn3** [comment]
I hadnt tested any of them yet. Thanks for testing `bingrep` though.

**Comment by jyn514** [comment]
`ripgrep` fails with the same `faerie` error about addends in relocations. `bgrep` gives a whole bunch of 'undefined reference' errors at link time:

```
Compiling bgrep v1.0.0
[codegen mono items] start
[codegen mono items] end time: 6.034330509s
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/joshua/src/rust/rustc_codegen_cranelift/build_sysroot/sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/tmp/cargo-installgCjZJh/x86_64-unknown-linux-gnu/release/deps/bgrep-95e1d2faa1ae6220.some_file.o.rcgu.o" "-o" "/tmp/cargo-installgCjZJh/x86_64-unknown-linux-gnu/release/deps/bgrep-95e1d2faa1ae6220" "/tmp/cargo-installgCjZJh/x86_64-unknown-linux-gnu/release/deps/bgrep-95e1d2faa1ae6220.allocator_shim.o.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/tmp/cargo-installgCjZJh/x86_64-unknown-linux-gnu/release/deps" "-L" "/tmp/cargo-installgCjZJh/release/deps" "-L" "/home/joshua/src/rust/rustc_codegen_cranelift/build_sysroot/sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/tmp/rustcwOjofE/libbacktrace_sys-6dbde68fbe8a4381.rlib" "-Wl,--end-group" "/home/joshua/Documents/Programming/rust/rustc_codegen_cranelift/build_sysroot/sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-247a06f7027ddb1d.rlib" "-Wl,-Bdynamic" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
  = note: /usr/local/bin/ld: /tmp/cargo-installgCjZJh/x86_64-unknown-linux-gnu/release/deps/bgrep-95e1d2faa1ae6220.some_file.o.rcgu.o: in function `core::option::Option<T>::unwrap':
          /home/joshua/Documents/Programming/rust/rustc_codegen_cranelift/build_sysroot/sysroot_src/src/libcore/macros.rs:12: undefined reference to `core::panicking::panic'
... truncated ...
/usr/local/bin/ld: /tmp/cargo-installgCjZJh/x86_64-unknown-linux-gnu/release/deps/bgrep-95e1d2faa1ae6220.allocator_shim.o.rcgu.o: in function `__rust_alloc_zeroed':
          allocator_shim.o:(.text.__rust_alloc_zeroed+0x6): undefined reference to `__rdl_alloc_zeroed'
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

error: failed to compile `bgrep v1.0.0`, intermediate artifacts can be found at `/tmp/cargo-installgCjZJh
````

**Comment by bjorn3** [comment]
I can't reproduce the `bgrep` link error. It compiles and runs fine for me with `rustc 1.40.0-nightly (2748a9fd9 2019-10-08)`:

```
$ ../cargo.sh run -- -b bgrep target/x86_64-unknown-linux-gnu/debug/bgrep
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/x86_64-unknown-linux-gnu/debug/bgrep -ob bgrep target/x86_64-unknown-linux-gnu/debug/bgrep -h`
0x632be3
0x632c0d
0x632c7b
0x632ca5
[...]
```

**Comment by bjorn3** [comment]
I am able to reproduce it by building `bgrep` in release mode. I will see if I can fix it.

For now you could use `./cargo.sh install --debug bgrep`. `cg_clif` doesn't perform any extra optimizations in release mode anyway to workaround https://github.com/CraneStation/cranelift/issues/732.

**Comment by jyn514** [comment]
Yup installing in debug mode made it work.

**Comment by bjorn3** [comment]
Opened https://github.com/bjorn3/rustc_codegen_cranelift/issues/746 for this.

**Comment by bjorn3** [comment]
* [x] https://github.com/leod/rendology (at least on the thread branch)

**Comment by bjorn3** [comment]
Pretty much all applications mentioned here now work. Only `zemeroth` doesn't work yet, because of some trouble with `c2-chacha` patching.

**Comment by bjorn3** [comment]
Going to close this as I don't think I have seen a misconpilation in quite a while. Pretty much all programs cg_clif can't compile right now are due to missing features rather than bugs in cg_clif. If anyone finds a program cg_clif can't handle please do report it as a new issue.
